components/builder-api-client/src/lib.rs

```diff
@@ -1104,21 +1104,35 @@ impl Client {
     /// * Remote depot unavailable
     pub fn search_package(&self,
                           search_term: &str,
+                          limit: usize,
                           token: Option<&str>)
-                          -> Result<(Vec<PackageIdent>, bool)> {
-        let mut res = self.maybe_add_authz(self.0.get(&package_search(search_term)), token)
-                          .send()?;
-        match res.status {
-            StatusCode::Ok | StatusCode::PartialContent => {
-                let mut encoded = String::new();
-                res.read_to_string(&mut encoded)
-                   .map_err(Error::BadResponseBody)?;
-                let package_results: PackageResults<PackageIdent> = serde_json::from_str(&encoded)?;
-                let packages: Vec<PackageIdent> = package_results.data;
-                Ok((packages, res.status == StatusCode::PartialContent))
-            }
-            _ => Err(err_from_response(res)),
+                          -> Result<(Vec<PackageIdent>, isize)> {
+        let mut packages = Vec::new();
+        let mut total = 0;
+        let mut status_code = StatusCode::PartialContent;
+        while packages.len() < limit && status_code == StatusCode::PartialContent {
+            let req = self.0
+                          .get_with_custom_url(&package_search(search_term), |url| {
+                              url.set_query(Some(&format!("range={:?}&distinct=true",
+                                                          packages.len())));
+                          });
+            let mut res = self.maybe_add_authz(req, token).send()?;
+            match res.status {
+                StatusCode::Ok | StatusCode::PartialContent => {
+                    let mut encoded = String::new();
+                    res.read_to_string(&mut encoded)
+                       .map_err(Error::BadResponseBody)?;
+                    let package_results: PackageResults<PackageIdent> =
+                        serde_json::from_str(&encoded)?;
+                    packages.extend(package_results.data);
+                    total = package_results.total_count;
+                    status_code = res.status;
+                }
+                _ => return Err(err_from_response(res)),
+            };
         }
+        packages.truncate(limit);
+        Ok((packages, total))
```

Comment:
In order to avoid the nonsensical `isize` leaking out to callers, I think it would be better to clean thing up here (if we can't change `PackageResults` itself):
```suggestion
        total = if package_results.total_count > 0 {
           packages.len().max(package_results.total_count as usize)
        } else {
            packages.len()
        };
        Ok((packages, total))
```

Reply:
The isize is due to PG/Diesel.. I _think_ it might be fine to change PackageResult itself.. but it's probably safest to just adjust the return here to be usize.  It should also be safe to cast to usize as we're not actually ever going to return negative counts.